nilearn/input_data/nifti_spheres_masker.py

```diff
@@ -0,0 +1,219 @@
+"""
+Transformer for computing seeds signals.
+"""
+import warnings
+
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.externals.joblib import Memory
+
+from .. import _utils
+from .._utils import logger, CacheMixin
+from .._utils.niimg_conversions import check_niimg
+from .. import signal
+from .. import image
+from .. import masking
+
+
+def _signals_from_seeds(seeds, niimg, radius=None, mask_img=None):
+    """ Note: this function is sub-optimal for voxel size radius
+    """
+    n_seeds = len(seeds)
+    niimg = check_niimg(niimg)
+    shape = niimg.get_data().shape
+    affine = niimg.get_affine()
+    if mask_img is not None:
+        mask_img = check_niimg(mask_img, ensure_3d=True)
+        mask_img = image.resample_img(mask_img, target_affine=affine,
+                                      target_shape=shape[:3],
+                                      interpolation='nearest')
+        mask, _ = masking._load_mask_img(mask_img)
+    signals = np.empty((shape[3], n_seeds))
+    # Create an array of shape (3, array.shape) containing the i, j, k indices
+    # in voxel space
+    coords = np.vstack((np.indices(shape[:3]),
+                        np.ones((1,) + shape[:3])))
+    # Transform the indices into native space
+    coords = np.tensordot(affine, coords, axes=[[1], [0]])[:3]
+    # Matrix where we stor each sphere to check overlaps
+    mark_map = - np.ones(shape[:3])
+
+    for i, seed in enumerate(seeds):
+        seed = np.asarray(seed)
+        # Compute square distance to the seed
+        dist = ((coords - seed[:, None, None, None]) ** 2).sum(axis=0)
+        if radius is None or radius ** 2 < np.min(dist):
+            dist_mask = (dist == np.min(dist))
+        else:
+            dist_mask = (dist <= radius ** 2)
+        if mask_img is not None:
+            dist_mask = np.logical_and(mask, dist_mask)
+        if not dist_mask.any():
+            raise ValueError('Seed #%i is out of the mask' % i)
+        # Check if there is overlapping with previous seeds
+        markers = np.unique(mark_map[dist_mask])
+        if len(markers) != 1:
+            warnings.warn('Sphere #%i overlaps with sphere #%i'
+                          % (i, markers[1]))
+        signals[:, i] = np.mean(niimg.get_data()[dist_mask], axis=0)
+    return signals
+
+
+class NiftiSpheresMasker(BaseEstimator, TransformerMixin, CacheMixin):
+    """Class for masking of Niimg-like objects using seeds.
+
+    NiftiSpheresMasker is useful when data from given seeds should be
+    extracted. Use case: Summarize brain signals from seeds that were
+    obtained from prior knowledge.
+
+    Parameters
+    ==========
+    seeds: List of triplet of coordinates in native space
+        Seed definitions. List of coordinates of the seeds in the same space
+        as the images (typically MNI or TAL).
+
+    radius: float, optional.
```

Comment:
2 inconsistencies in this docstring
a) point at the end of variable type
b) point at the end of variable description